### PR TITLE
Add a tutorial on how to use aggregations with OpenSearch

### PR DESCRIPTION
**Pull request by anelook**
New tutorial will help our users create and run aggregations with OpenSearch.
This is logical continuation of the previous tutorial - https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html

### CODE REVIEW

**Comment by lornajane** [comment]
Does this tutorial need to be a sequel to the previous one, or could it be useful standalone to someone with a bit of OpenSearch experience but not much with the aggregations aspect?

**Comment by anelook** [comment]
> Does this tutorial need to be a sequel to the previous one, or could it be useful standalone to someone with a bit of OpenSearch experience but not much with the aggregations aspect?

Since I didn't want to repeat the instructions for playground setup, and because we analyse same dataset I positioned it as a sequel. 

An alternative I thought of would be to split the first tutorial into preparation step and the actual search experiments. Then those parts can be referenced independently.

**Review by lornajane — COMMENTED** [review]
I picked up a few minor things and I have some high level comments - I'd like to review your next set of changes and do a language review as well before I approve. I think this is in good shape and the sample data set continues to give us some great examples. The beginning is a bit of a rough experience. Perhaps we can number the setup steps in the other article, and then say in this one to follow the first four (or however many) steps.  Alternatively, be clearer that the repository is required as the basis of the exercise.

File: docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+**********************
+
+In the :doc:`previous tutorial <opensearch-and-nodejs>` we've learned how to setup an OpenSearch cluster and its JavaScript client, how to index data and retrieve a dynamically created schema. Use this knowledge to prepare your playground for aggregations:
+
+You'll need an OpenSearch cluster and a NodeJS project. Make sure that you have `the necessary essentials <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#prerequisites>`_ in place, that you're `connected to the OpenSearch cluster <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#connect-to-opensearch>`_ and have `example data already ingested <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#load-example-recipe-data-and-ingest-it-into-the-cluster>`_.
+
+You can also clone the final demo project from `GitHub repository <https://github.com/aiven/demo-open-search-node-js>`_. You'll find the code for aggregations in the `aggregate.js <https://github.com/aiven/demo-open-search-node-js/blob/main/aggregate.js>`_.
+
+
+About aggregations
```
[the remark this answers is not in the document]
Comment (anelook) [reply docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst]: Done: moved to a separate concept article.

File: docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+          size: 0, // we're not interested in `hits`
+        }
+      );
+
+
+The best way to learn more about each type of aggregations is to try them out. Therefore, it's time to make our hands dirty and do some coding. Create `aggregate.js` file, this is where we'll be adding our code. At the top of the file import client and index name, we'll need them to send requests to the cluster.
+
+.. code-block:: javascript
+
+    const { client, indexName: index } = require("./config");
```
[the remark this answers is not in the document]
Comment (anelook) [reply docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:56]: I've introduced different files in the step of preparation of a playground. This hopefully helps.

File: docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+
+Prepare the playground
+**********************
+
+In the :doc:`previous tutorial <opensearch-and-nodejs>` we've learned how to setup an OpenSearch cluster and its JavaScript client, how to index data and retrieve a dynamically created schema. Use this knowledge to prepare your playground for aggregations:
+
+You'll need an OpenSearch cluster and a NodeJS project. Make sure that you have `the necessary essentials <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#prerequisites>`_ in place, that you're `connected to the OpenSearch cluster <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#connect-to-opensearch>`_ and have `example data already ingested <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#load-example-recipe-data-and-ingest-it-into-the-cluster>`_.
+
+You can also clone the final demo project from `GitHub repository <https://github.com/aiven/demo-open-search-node-js>`_. You'll find the code for aggregations in the `aggregate.js <https://github.com/aiven/demo-open-search-node-js/blob/main/aggregate.js>`_.
```
[the remark this answers is not in the document]
Comment (anelook) [reply docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst]: Added proper details, so no step is missing. Playground setup is now a separate tutorial

File: docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+     - (0 + 3 + 4) / 3 = 2.3(3)
+   * - 2000
+     - 0
+     - (3 + 4 + 0) / 3 = 2.3(3)
+   * - ...
+     - ...
+     - and so on
+
+
+For every data point (a year in our case) we take the count of added recipes, add number of recipes added over last two years and divide the result by three (according to the size of our window). For the first and second year we divide by the number of available years (1 and 2 respectively). And this is how moving average is calculated. If you compare numbers from the table with the numbers returned in the `moving_average` field of the response body, you can see they are same.
```
[the remark this answers is not in the document]
Comment (anelook) [reply docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst:748]: Happy that you could easily follow it! Pipeline aggregations are very tricky to explain. And recipe data isn't perfect for them. I want to write a separate article and use some log data instead.

File: docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+          size: 0,
+        },
+        logAggs.bind(this, `rare-terms-aggs-for-${field}`)
+      );
+    };
+
+
+::
+
+    run-func aggregate terms categories.keyword 3
```
[the remark this answers is not in the document]
Comment (anelook) [reply docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst]: Hehe, not sure how it sneaked in! Fixed

File: docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+        '95.0': 1317.4926233766223,
+        '99.0': 3256.4999999999945
+      }
+    }
+
+From the returned result you can see that 50% of recipes have less than 331 calories. Interestingly, only one percent of the meals is more than 3256 calories. You must be curious what falls within that last percentile ;) Now that we know the value to look for, we can use `a range query <https://developer.aiven.io/docs/products/opensearch/howto/opensearch-and-nodejs.html#find-fields-with-a-value-within-a-range>`_ to find the recipes:
+
+::
+
+    run-func index rangeSearch calories 3256
```
[the remark this answers is not in the document]
Comment (anelook) [reply docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.rst]: It can be unbounded, I added extra phrase about it

**Comment by anelook** [comment]
Based on the feedback I revisited unclear chunks and made these improvements:

- Moved playground setup to a separate article, which is referenced by both Search and Aggregations tutorials.
- Moved description of aggregations into concept article
- Replaced Present Perfect tense with Past tense
- Clarified what code goes to what files. Instead of keeping all in a single index.js, introduced config.js, index.js, helpers.js, search.js and aggregation.js and explained their roles.
- Minor simplifications and alignment when it comes to code and commands.

**Comment by anelook** [comment]
Direct links to added and modified content

Setup Playground: https://deploy-preview-310--developer-aiven-io-preview.netlify.app/docs/products/opensearch/howto/setup-opensearch-and-nodejs-playground.html

Search: https://deploy-preview-310--developer-aiven-io-preview.netlify.app/docs/products/opensearch/howto/opensearch-and-nodejs.html

Aggregations: https://deploy-preview-310--developer-aiven-io-preview.netlify.app/docs/products/opensearch/howto/opensearch-aggregations-and-nodejs.html

Aggregation concepts: http://127.0.0.1:8000/docs/products/opensearch/concepts/aggregations.html

**Comment by lornajane** [comment]
Hmm, I sort of like this but I think there are still some really useful pieces of documentation hiding in this very specific set of tutorials ... trying to think about the right structure for this is making my head hurt! There's a lot going on here, and I think we need to move some of these things out into reference articles in their own right that can be used by anyone to do these things for their own projects rather than in a grouped playground setup task.
 - creating an opensearch service. This should be in getting started, if it isn't we should fix that, and just link to it.
 - importing the data. We already have a sample data setup page. Can we direct people to this existing resource? Even if we add the Node equivalent instructions after the Python ones.
 - initialising a node project. We aren't teaching NodeJS so let's remove this section (but the overview of the files is useful to orientate for the search and aggregate articles)
 - basic use of `avn service` commands - I think this is out of scope when teaching someone about aggregations but it would be fine to add a tip linking to the CLI docs
 - Connecting to NodeJS - this should be a code sample, people will want this outside of these specific articles.

I think the PostgreSQL section is a good illustration of how I was thinking about this content. There's a post about getting started, one about sample data, some connection examples, and then some other pieces that refer back to these "building blocks" - but each of the blocks are standalone and discoverable.

**Comment by anelook** [comment]
Added changes after our conversation about the structure.

**Comment by lornajane** [comment]
@anelook This is in good shape! I've added some language edits that you might want to double-check, and there's a small conflict.

**Comment by anelook** [comment]
Thank you for the language improvements! All looks good to me. 

I've pushed a rebased version and GitHub demands a new review.

**Comment by anelook** [comment]
I've rebased the branch, it is up to date in relation to ``main``.

File: _toc.yml [hunk 7]
```diff
@@ -59,6 +60,7 @@ entries:
         title: HowTo
         entries:
           - file: docs/products/kafka/howto/kafka-sasl-auth
+          - file: docs/products/kafka/howto/integrate-service-logs-into-kafka-topic
```
Comment (lornajane) [inline _toc.yml:63]: I'm not sure what this change is doing here!

Comment (anelook) [reply _toc.yml]: 😨 This was not supposed to be there! Thx for catching it. Removed

**Review by lornajane — APPROVED** [review]
Great job, I know it was a long road!

**Pull request merged**